[Parse/Syntax] Refactoring to decouple the parser from syntax tree creation

akyrtzi opened the pull request:
Instead of creating syntax nodes directly, modify the parser to invoke an abstract interface 'SyntaxParseActions' while it is parsing the source code.
This decouples the act of parsing from the act of forming a syntax tree representation.
'SyntaxTreeCreator' is an implementation of SyntaxParseActions that handles the logic of creating a syntax tree.
To enforce the layering separation of parsing and syntax tree creation, a static library swiftSyntaxParse is introduced to compose the two.

This decoupling is important for introducing a syntax parser library for SwiftSyntax to directly access parsing.

akyrtzi committed:
[Parse] Add a guard macro for 'SyntaxParserResult.h'

akyrtzi committed:
[Parse/Syntax] Refactoring to decouple the parser from syntax tree creation

Instead of creating syntax nodes directly, modify the parser to invoke an abstract interface 'SyntaxParseActions' while it is parsing the source code.
This decouples the act of parsing from the act of forming a syntax tree representation.
'SyntaxTreeCreator' is an implementation of SyntaxParseActions that handles the logic of creating a syntax tree.
To enforce the layering separation of parsing and syntax tree creation, a static library swiftSyntaxParse is introduced to compose the two.

This decoupling is important for introducing a syntax parser library for SwiftSyntax to directly access parsing.

akyrtzi commented:
@swift-ci smoke test

akyrtzi committed:
[utils/gyb_syntax_support] Fix python-lint issues

akyrtzi commented:
@swift-ci smoke test

akyrtzi committed:
[ParsedRawSyntaxNode] Make sure move-assignment operator calls the move-constructor

akyrtzi commented:
@swift-ci smoke test

akyrtzi committed:
[SyntaxTreeCreator] Add documentation comments for the SyntaxTreeCreator class

akyrtzi commented:
I just realized that doing a direct `ParsedSyntaxRecorder::record[some syntax]` call from the parser is not a good idea due to possibility of being in a backtracking context when the call is made. I'm going to push a change to change the calls to `ParsedSyntaxRecorder::make[some syntax]` which will implicitly check for backtracking and create a `recorded` or `deferred` node accordingly.

akyrtzi committed:
[Parse/Syntax] Replace ParsedSyntaxRecorder::record* calls from the parser with ParsedSyntaxRecorder::make*

Doing a "direct ParsedSyntaxRecorder::record[some syntax]" call from the parser is not a good idea due to possibility
of being in a backtracking context when the call is made. Replace them with "ParsedSyntaxRecorder::make[some syntax]"
which will implicitly check for backtracking and create a recorded or deferred node accordingly.

rintaro commented:
@swift-ci Please ASAN test

rintaro commented on lib/Parse/SyntaxParsingContext.cpp:
```diff
@@ -157,42 +166,49 @@ const SyntaxParsingContext *SyntaxParsingContext::getRoot() const {
-      Arena, Tok.getKind(), Text, LeadingTrivia.Pieces, TrailingTrivia.Pieces));
+  ParsedRawSyntaxNode raw;
+  if (IsBacktracking)
+    raw = ParsedRawSyntaxNode::makeDeferred(Tok, LeadingTrivia, TrailingTrivia);
```
IIRC, current `Lexer` creates `TriviaPiece` with refcounted owned text. It would be nice if `Lexer` creates source referencing `TriviaPiece`s, then `SyntaxTreeCreator::recordToken()` turns them into "owned".

akyrtzi replied:
I agree, and actually my intention is to make changes for the `Lexer` to only determine the source ranges for leading and trailing trivia, and only do the processing to determine the actual trivia pieces at `record` time, in `ParsedRawSyntaxRecorder`. That will prevent re-processing the same trivia pieces when doing backtracking.

However, I'd like to hold off on performance optimizations until after SwiftSyntax adopts the new model.

rintaro commented on include/swift/Parse/ParsedRawSyntaxRecorder.h:
```diff
@@ -0,0 +1,72 @@
+                                  const syntax::Trivia &leadingTrivia,
+                                  const syntax::Trivia &trailingTrivia);
+
+  /// Record a missing token. \c loc can be invalid or an approximate location
```
Nitpick: You should use `\p` to annotate function arguments.

akyrtzi replied:
Done.

rintaro commented on include/swift/Parse/SyntaxParseActions.h:
```diff
@@ -0,0 +1,68 @@
+
+  /// Used for incremental re-parsing.
+  virtual std::pair<size_t, OpaqueSyntaxNode>
+      lookupNode(size_t lexerOffset, syntax::SyntaxKind kind) {
```
I think 
```cpp
  virtual std::pair<size_t, OpaqueSyntaxNode> 
  lookupNode(size_t lexerOffset, syntax::SyntaxKind kind) {

  }
```
is the standard indentation.

akyrtzi replied:
Done.

rintaro commented on include/swift/SyntaxParse/SyntaxTreeCreator.h:
```diff
@@ -0,0 +1,70 @@
+                                   CharSourceRange range) override;
+
+  std::pair<size_t, OpaqueSyntaxNode>
+      lookupNode(size_t lexerOffset, syntax::SyntaxKind kind) override;
```
ditto

akyrtzi replied:
Done.

jrose-apple commented:
What's the impact on parse time?

akyrtzi commented:
> What's the impact on parse time?

There's no impact on parse time for the compiler because the libSyntax parsing code path is completely disabled (and I've made no changes to the `Lexer`).

For libSyntax C++ tree creation itself I didn't bother doing measurements because the architectural change for SwiftSyntax will obviate the need to create the C++ syntax tree for SwiftSyntax and make SwiftSyntax multiple times faster.

However there are some performance optimizations to make to speed-up the libSyntax parsing path, which I will push for after we adopt the new model for SwiftSyntax.

akyrtzi commented:
@swift-ci Please ASAN test

nkcsgexi commented on include/swift/Parse/ParsedRawSyntaxNode.h:
```diff
@@ -0,0 +1,252 @@
+  void releaseMemory() {
+    switch (DK) {
+    case DataKind::Recorded:
+      RecordedData.~RecordedSyntaxNode(); break;
```
Nitpick: `break;` on separate lines.

akyrtzi replied:
Done.

nkcsgexi commented on include/swift/Parse/ParsedRawSyntaxRecorder.h:
```diff
@@ -0,0 +1,72 @@
+  /// \c kind. Missing optional elements are represented with a null
+  /// ParsedRawSyntaxNode object.
+  ParsedRawSyntaxNode recordExactRawSyntax(syntax::SyntaxKind kind,
+                                     ArrayRef<ParsedRawSyntaxNode> elements);
```
`recordRawSyntax` is good enough IMO. We don't have a non-exact version of recording and it doesn't make much sense.

akyrtzi replied:
Done.

jrose-apple commented:
> > What's the impact on parse time?
> 
> There's no impact on parse time for the compiler because the libSyntax parsing code path is completely disabled (and I've made no changes to the `Lexer`).

We enable it for tests, at least, don't we?

nkcsgexi commented on include/swift/Parse/ParsedRawSyntaxNode.h:
```diff
@@ -0,0 +1,252 @@
+/// in the current parsing context.
+class ParsedRawSyntaxNode {
+  enum class DataKind: uint8_t {
+    Recorded,
```
could you add some comments why we don't need `RecordedLayout` and `RecordedToken`?

akyrtzi replied:
Done.

akyrtzi commented:
> We enable it for tests, at least, don't we?

Good point. I measured and it's ~15% regression in parse time for C++ libSyntax. In absolute terms this is +6.5 ms for a 4,120 LOC file.

jrose-apple commented:
That is non-wonderful but probably livable, given that parse time remains a small percentage of overall time for fully-compiled tests.

akyrtzi committed:
Address PR feedback

* Add and improve documentation comments
* Adjust formatting
* Rename recordExactRawSyntax -> rec.recordRawSyntax

akyrtzi commented on include/swift/Parse/SyntaxParsingContext.h:
```diff
@@ -108,6 +42,11 @@ enum class SyntaxContextKind {
+enum class SyntaxNodeCreationKind {
+  Recorded,
+  Deferred,
+};
```
Done.

nkcsgexi replied:
Minor nitpick: will it be better to name as "finalized" vs. "pending".

akyrtzi committed:
[Parse/ParsedRawSyntaxRecorder] In recordToken() adjust the range to account for backticks in an identifier

rintaro commented on lib/Parse/ParsedRawSyntaxNode.cpp:
```diff
@@ -0,0 +1,67 @@
+  dumpSyntaxKind(OS, getKind());
+
+  if (isToken()) {
+    OS << ' ' << getTokenText(getTokenKind());
```
This crashes if *not* `isTokenTextDetermined(getTokenKind())`

akyrtzi replied:
Fixed, thanks!

akyrtzi committed:
[Parse] Ensure that the tokens formed for interpolation anchors have proper ranges from the input source

akyrtzi committed:
[ParsedRawSyntaxNode] Fix ParsedRawSyntaxNode::dump()

Using dumpTokenKind() function instead of getTokenText().

akyrtzi committed:
[ParsedRawSyntaxNode] Explicitely specify if a ParsedRawSyntaxNode is null using a DataKind

Avoid implicitely assuming 'null' node if its OpaqueSyntaxNode is null, there should be no interpretation
of OpaqueSyntaxNode values, a SyntaxParseActions implementation should be able to return null pointers as OpaqueSyntaxNode.

rintaro commented on include/swift/Parse/CMakeLists.txt:
```diff
@@ -0,0 +1,16 @@
+  ParsedSyntaxRecorder.h.gyb
+)
+
+add_gyb_target(swift-parse-syntax-generated-headers
```
I think it's safe to add this target to `LLVM_COMMON_DEPENDS` to ensure they are generated prior to any compilation of cpp files. like:
 https://github.com/apple/swift/blob/c932f60403edef804d38b4c29ae99d59ef5eb62b/lib/CMakeLists.txt#L13-L14

akyrtzi replied:
You have a good point, since those headers are "globally" accessed. I believe we can move all the `Parsed*` and `SyntaxParsingContext.h` headers in `lib/Parse` but until this happens the gyb target should be added to `LLVM_COMMON_DEPENDS` as you suggested.

nkcsgexi reviewed (approved):
No more comments from me 👍

akyrtzi committed:
[CMake] Add the gyb target 'swift-parse-syntax-generated-headers' to LLVM_COMMON_DEPENDS

Since the generated headers are in 'include/swift/Parse' they are globally accessible so make sure they are created early on.
If these get moved inside 'lib/Parse' then we can remove this dependency setting.

akyrtzi commented:
@swift-ci smoke test
